src/algebra/module.lean

```diff
@@ -99,17 +97,68 @@ lemma finset.sum_smul {f : ι → R} {s : finset ι} {x : M} :
   s.sum f • x = s.sum (λ r, (f r) • x) :=
 ((smul_add_hom R M).flip x).map_sum f s
 
-end semimodule
+end add_comm_monoid
 
-section prio
-set_option default_priority 100 -- see Note [default priority]
-/-- A module is a generalization of vector spaces to a scalar ring.
-  It consists of a scalar ring `R` and an additive group of "vectors" `M`,
-  connected by a "scalar multiplication" operation `r • x : M`
-  (where `r : R` and `x : M`) with some natural associativity and
-  distributivity axioms similar to those on a ring. -/
-class module (R : Type u) (M : Type v) [ring R] [add_comm_group M] extends semimodule R M
-end prio
+section add_comm_group
+
+variables (R M) [semiring R] [add_comm_group M]
+
+/-- A structure containing most informations as in a semimodule, except the fields `zero_smul`
+and `smul_zero`. As these fields can be deduced from the other ones when `M` is an `add_comm_group`,
+this provides a way to construct a semimodule structure by checking less properties, in
+`semimodule.of_core`. -/
+@[nolint has_inhabited_instance]
+structure semimodule.core extends has_scalar R M :=
+(smul_add : ∀(r : R) (x y : M), r • (x + y) = r • x + r • y)
+(add_smul : ∀(r s : R) (x : M), (r + s) • x = r • x + s • x)
+(mul_smul : ∀(r s : R) (x : M), (r * s) • x = r • s • x)
+(one_smul : ∀x : M, (1 : R) • x = x)
+
+variables {R M}
+
+/-- Define `semimodule` without proving `zero_smul` and `smul_zero` by using an auxiliary
+structure `semimodule.core`, when the underlying space is an `add_comm_group`. -/
+def semimodule.of_core (H : semimodule.core R M) : semimodule R M :=
+by letI := H.to_has_scalar; exact
+{ zero_smul := λ x,
+    have (0 : R) • x + (0 : R) • x = (0 : R) • x + 0, by rw ← H.add_smul; simp,
+    add_left_cancel this,
+  smul_zero := λ r,
+    have r • (0:M) + r • 0 = r • 0 + 0, by rw ← H.smul_add; simp,
+    add_left_cancel this,
+  ..H }
+
+variable [semimodule R M]
+
+@[simp] theorem smul_neg (r : R) (x : M) : r • (-x) = -(r • x) :=
+eq_neg_of_add_eq_zero (by simp [← smul_add])
+
+theorem smul_sub (r : R) (x y : M) : r • (x - y) = r • x - r • y :=
+by simp [smul_add, sub_eq_add_neg]; rw smul_neg
+
+end add_comm_group
+
+/--
+Modules are defined as an `abbreviation` for semimodules,
+if the base semiring is a ring.
+(A previous definition made `module` a structure
+defined to be `semimodule`.)
+This has as advantage that modules are completely transparent
+for type class inference, which means that all instances for semimodules
+are immediately picked up for modules as well.
+A cosmetic disadvantage is that one can not extend modules an sich,
```

Comment:
```suggestion
A cosmetic disadvantage is that one can not extend modules as such,
```

Reply:
What's wrong with German?

Reply:
I copied this from the docstring on vector spaces. Since this is meaningful German, I thought it was a classy and accepted way to say `by itself` in English (in the same way that you can sometimes put French words in an English sentence, like `déjà-vu`). If it's not standard, I will change it (at both places). Tell me!

Reply:
I can't remember seeing that used in English text, but we Australians are pretty ignorant about this other-languages stuff. :-)

Reply:
Changed.